src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

```diff
@@ -25,56 +25,54 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
                 return null;
             }
 
-            var descriptor = JObject.Load(reader);
-            var descriptorKind = descriptor[nameof(TagHelperDescriptor.Kind)].Value<string>();
-            var typeName = descriptor[nameof(TagHelperDescriptor.Name)].Value<string>();
-            var assemblyName = descriptor[nameof(TagHelperDescriptor.AssemblyName)].Value<string>();
-            var tagMatchingRules = descriptor[nameof(TagHelperDescriptor.TagMatchingRules)].Value<JArray>();
-            var boundAttributes = descriptor[nameof(TagHelperDescriptor.BoundAttributes)].Value<JArray>();
-            var childTags = descriptor[nameof(TagHelperDescriptor.AllowedChildTags)].Value<JArray>();
-            var documentation = descriptor[nameof(TagHelperDescriptor.Documentation)].Value<string>();
-            var tagOutputHint = descriptor[nameof(TagHelperDescriptor.TagOutputHint)].Value<string>();
-            var caseSensitive = descriptor[nameof(TagHelperDescriptor.CaseSensitive)].Value<bool>();
-            var diagnostics = descriptor[nameof(TagHelperDescriptor.Diagnostics)].Value<JArray>();
-            var metadata = descriptor[nameof(TagHelperDescriptor.Metadata)].Value<JObject>();
-
+            // Required tokens (order matters)
+            var descriptorKind = ReadNextStringProperty(reader, nameof(TagHelperDescriptor.Kind));
+            var typeName = ReadNextStringProperty(reader, nameof(TagHelperDescriptor.Name));
+            var assemblyName = ReadNextStringProperty(reader, nameof(TagHelperDescriptor.AssemblyName));
             var builder = TagHelperDescriptorBuilder.Create(descriptorKind, typeName, assemblyName);
 
-            builder.Documentation = documentation;
-            builder.TagOutputHint = tagOutputHint;
-            builder.CaseSensitive = caseSensitive;
-
-            foreach (var tagMatchingRule in tagMatchingRules)
-            {
-                var rule = tagMatchingRule.Value<JObject>();
-                builder.TagMatchingRule(b => ReadTagMatchingRule(b, rule, serializer));
-            }
-
-            foreach (var boundAttribute in boundAttributes)
+            ReadProperties(reader, propertyName =>
```

Comment:
I found that by inlining every call to `ReadProperties` it saved 12% additional allocations (no lambda closures); however, it did not improve op/s shockingly. I felt that it was a drastic enough change that reduced the maintainability that the 12% reduction for 0 op/s improvement wasn't worth it. It's something we can always re-evaluate later if these optimizations prove to not be enough